validations/src/validations.rs

```diff
@@ -3569,3 +3388,665 @@ mod tests {
+    #[allow(clippy::cast_possible_truncation)]
+    #[test]
+    fn reppoe_overflow_v1() {
+        // Test the behavior of the algorithm when our node has 0 reputation
+        let mut rep_engine = ReputationEngine::new(1000);
+        let id0 = PublicKeyHash::from_bytes(&[0; 20]).unwrap();
+        let id1 = PublicKeyHash::from_bytes(&[1; 20]).unwrap();
+        rep_engine.ars_mut().push_activity(vec![id0]);
+        rep_engine.ars_mut().push_activity(vec![id1]);
+        rep_engine
+            .trs_mut()
+            .gain(Alpha(10), vec![(id0, Reputation(u32::max_value() - 2))])
+            .unwrap();
+
+        // Test big values that result in < 100%
+        let (t01, p01) = calculate_reppoe_threshold_v1(&rep_engine, &id0, 1, 2000);
+        assert_eq!(t01, Hash::with_first_u32(0xFFFF_FFFE));
+        assert_eq!((p01 * 100_f64).round() as i128, 100);
+    }
+
+    // FIXME: Allow for now, wait for https://github.com/rust-lang/rust/issues/67058 to reach stable
+    #[allow(clippy::cast_possible_truncation)]
+    #[test]
+    fn reppoe_overflow_v2() {
+        // Test the behavior of the algorithm when our node has 0 reputation
+        let mut rep_engine = ReputationEngine::new(1000);
+        let id0 = PublicKeyHash::from_bytes(&[0; 20]).unwrap();
+        let id1 = PublicKeyHash::from_bytes(&[1; 20]).unwrap();
+        rep_engine.ars_mut().push_activity(vec![id0]);
+        rep_engine.ars_mut().push_activity(vec![id1]);
+        rep_engine
+            .trs_mut()
+            .gain(Alpha(10), vec![(id0, Reputation(u32::max_value() - 2))])
+            .unwrap();
+
+        // Test big values that result in < 100%
+        let (t01, p01) = calculate_reppoe_threshold_v2(&rep_engine, &id0, 1, 2000);
+        assert_eq!(t01, Hash::with_first_u32(0xFFFF_FFFF / 2000));
+        // 1 / 2000 = 500 / 1_000_000
+        assert_eq!((p01 * 1_000_000_f64).round() as i128, 500);
+    }
+
+    // FIXME: Allow for now, wait for https://github.com/rust-lang/rust/issues/67058 to reach stable
+    #[allow(clippy::cast_possible_truncation)]
+    #[test]
+    fn reppoe_overflow_v3() {
+        // Test the behavior of the algorithm when our node has 0 reputation
+        let mut rep_engine = ReputationEngine::new(1000);
+        let id0 = PublicKeyHash::from_bytes(&[0; 20]).unwrap();
+        let id1 = PublicKeyHash::from_bytes(&[1; 20]).unwrap();
+        rep_engine.ars_mut().push_activity(vec![id0]);
+        rep_engine.ars_mut().push_activity(vec![id1]);
+        rep_engine
+            .trs_mut()
+            .gain(Alpha(10), vec![(id0, Reputation(u32::max_value() - 2))])
+            .unwrap();
+
+        // Test big values that result in < 100%
+        let (t01, p01) = calculate_reppoe_threshold_v3(&rep_engine, &id0, 1, 2000);
+        assert_eq!(t01, Hash::with_first_u32(0xFFFF_FFFE));
+        assert_eq!((p01 * 100_f64).round() as i128, 100);
+    }
+}
```

Comment:
This test was mind blowing for me because despite of being correct, it represents a network condition that is highly unlikely.

It is important to remember that the minimum difficulty is only activated in case of total active reputation is below 2000:

```
let dr_eligibity = identity_eligibility / max(total_active_reputation, minimum_difficulty) * replication_factor
```

This test shows a situation in which there is only 1 active identity with most of reputation. Therefore, the `identity_eligilibity` (aka "trapezoid") is not coming into play as there is no other active reputed identity to redistribute eligibility. IMHO, this case should be a big problem because it would be likely to happen only after several rollbacks. It will be automatically solved after a data request with a replicacion factor greater than 1.

It would be nice to have some other tests checking a situation with more than 1 active reputed identity.

Reply:
> IMHO, this case should be a big problem

I assume you mean "should not"

Actually, this test has 2 identities, but 1 of them has 0 reputation so it is not taken into account in the trapezoid. I will add more tests.